File: Source/Aladdin.Math/Math/IRing.cs
namespace Aladdin.Math 
{
    ////////////////////////////////////////////////////////////////////////////
    // Интерфейс кольца
    ////////////////////////////////////////////////////////////////////////////
    public interface IRing<E> : IGroupAdd<E>
    {
        // признак единичного элемента
        bool IsOne(E a); E One { get; }
    
        // умножение/возведение в квадрат
        E Product(E a, E b); E Sqr(E a); 
    
        // вычисление частного и остатка
        E Divide(E a, E b); E Remainder(E a, E b);
    
        // вычисление частного и остатка
        E[] DivideAndRemainder(E a, E b);

        // вычисление степени элемента
        E Power(E a, BigInteger e); 
    
        // произведение степеней элементов 
        E PowerProduct(E P, BigInteger a, E Q, BigInteger b); 
    
        // расширенный алгоритм Евклида
        E[] Euclid(E A, E B); E GCD(E A, E B);
    }
}

File: Source/Aladdin.Math.Java/src/aladdin/math/IRing.java
package aladdin.math;
import java.math.*;

////////////////////////////////////////////////////////////////////////////
// Интерфейс кольца
////////////////////////////////////////////////////////////////////////////
public interface IRing<E> extends IGroupAdd<E>
{
    // признак единичного элемента
    boolean isOne(E a); E one();
    
    // умножение/возведение в квадрат
    E product(E a, E b); E sqr(E a); 
    
    // вычисление частного и остатка
    E divide(E a, E b); E remainder(E a, E b);
    
    // вычисление частного и остатка
    E[] divideAndRemainder(E a, E b);

    // вычисление степени элемента
    E power(E a, BigInteger e); 
    
    // произведение степеней элементов 
    E power_product(E P, BigInteger a, E Q, BigInteger b); 
    
    // расширенный алгоритм Евклида
    E[] euclid(E A, E B); E gcd(E A, E B);
}
